velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -118,14 +190,55 @@ void HiveDataSink::createWriter(
   // Without explicitly setting flush policy, the default memory based flush
   // policy is used.
   auto writerParameters = getWriterParameters(partitionName);
-
-  writerInfo_.emplace_back(std::make_shared<HiveWriterInfo>(*writerParameters));
-
   auto writePath = fs::path(writerParameters->writeDirectory()) /
       writerParameters->writeFileName();
+
   auto sink = dwio::common::DataSink::create(writePath);
-  writers_.push_back(std::make_unique<Writer>(
-      options, std::move(sink), *connectorQueryCtx_->memoryPool()));
+  writers_[partitionId] = std::make_unique<Writer>(
+      options, std::move(sink), *connectorQueryCtx_->memoryPool());
+  writerInfo_[partitionId] =
+      std::make_shared<HiveWriterInfo>(*writerParameters);
+}
+
+void HiveDataSink::computePartitionRowCountsAndIndices(
+    uint64_t maxPartitionId,
+    memory::MemoryPool* pool) {
+  vector_size_t numPartitionIds = maxPartitionId + 1;
+  uint64_t numRows = partitionIds_.size();
+
+  partitionSizes_.resize(numPartitionIds);
+  std::fill(partitionSizes_.begin(), partitionSizes_.end(), 0);
+  partitionRows_.resize(numPartitionIds, nullptr);
+
+  if (partitionIdGenerator_->totalPartitions() == 1) {
+    partitionSizes_[maxPartitionId] = numRows;
+    return;
+  }
+
+  for (uint64_t id = 0; id < numPartitionIds; id++) {
+    if (partitionRows_[id] != nullptr &&
+        partitionRows_[id]->size() < numRows * sizeof(vector_size_t)) {
+      partitionRows_[id] = allocateIndices(numRows, pool);
+    }
+  }
+
+  for (vector_size_t row = 0; row < numRows; row++) {
+    uint64_t id = partitionIds_[row];
+
+    if (partitionRows_[id] == nullptr) {
+      partitionRows_[id] = allocateIndices(numRows, pool);
+    }
+
+    auto* rawIndex = partitionRows_[id]->asMutable<vector_size_t>();
```

Comment:
Consider storing raw pointers to these buffers to avoid casting for each row.

Reply:
Borrow it from your example